### DISCUSSION TITLE
Extending Config Contexts

### DISCUSSION THREAD

**Discussion by jamesditrapani** [peering-manager/peering-manager#505] [Ideas]
Opening a discussion around the best way to extend Config Contexts to further objects. 

At the moment only routing policies and router objects support config contexts where custom and adhoc data can be defined via JSON. Going forward we need to look at extending this to further objects or launching an inheritence system allowing users to define further adhoc info to different models.

The easiest way to do this currently is to simply bring Config Contexts to existing models such as IXP, IXP Session, BGP Group, Direct Session, ASN, etc. All of these objects are either exposed directly, or indirectly to the templating engine and can be accessed this way.

The other option drawn out by @gmazoyer is to define a multi stage config context where data can be defined per object type (e.g. all IXP models) and a config context for specific models (e.g. a single IXP), these would then be merged into the templating engine with with per model data taking presedence over per model type.

How are people currently using config contexts and what is going to ensure you benefit from them going forward? Do you currently have a road block when templating due to the current lack of config context support? Or do you simply have another idea how we can do this better? Drop a message below! 😃

**Comment by maznu** [comment]
I think adding Config Contexts to a variety of objects is really important, especially in the absence of a "custom fields" module like e.g. NetBox.  For example, there are a bunch of "BGP peer parameters" which have no place in Peering Manager currently — and I am sure there will be other vendor-specific things which could be storeed in Config Contexts:

* on direct sessions and peer groups and ASNs: `allowas-in` (I make no judgement about whether needing this is good network design)
* on direct sessions and peer groups and maybe even ASNs: `default-originate` (in line with PM's goal of being for all eBGP sessions, including downstream networks)
* on direct sessions and peer groups and ASNs: BGP `timers` (sure, you could specify default timers for all peering sessions in your template, but if a direct session happens to be someone running an underpowered router, e.g. MikroTik CCR1xxx, timers can trip due to CPU load)

Some of these it might make sense to have their own dedicated UI fields and values (e.g. `default-originate` maybe?). But I think many others would niche requirements for automation, and PM should leave it up to the netdevops teams how to use Config Contexts.

**Comment by maznu** [comment]
## tl;dr

Following [discussion about adding default-originate](https://github.com/peering-manager/peering-manager/issues/566) — where using tags was suggested — and [config contexts](https://github.com/peering-manager/peering-manager/issues/567), perhaps the "kitchen sink" should be the Tag model?

-----

Consider the concept of "originating a default route" is fairly simple, but implemented in different ways by different vendors:

* Cisco/etc treats it as a per-peer boolean (on/off) setting
* JunOS implements this as an export policy for a static default route injected into RIB (but not redistributed to FIB)
* MikroTik has a per-peer setting, but with tristate values `never`, `if-installed` and `always`

In other words, lots of NOS vendors give engineers a way to originate a default route as a special case.  And engineers will treat their favourite NOS' implementation of BGP as the canonical way to do BGP (I shall call this "vendorthink", something I'm guilty of too!).

Another "vendorthink" that can trip up users: several BGP configuration implementations do not allow multiple import/export policies for a peer or group (e.g. anything with a Cisco-like syntax, but also FRR, VyOS, EdgeOS, RouterOS, etc).  If I'm an engineer it might seem impossible to me how to implement a list of multiple routing policies as per Peering Manager.  We should signpost how PM's Cisco-mindset users can make a per-peer route-map that will `call` into each of the multiple policies defined in Peering Manager, thus achieving the same effect.

We should suggest some good practices for how users can build their policies in PM and how to make good templates.  For example, I don't like having special cases where the names of things (e.g. name of export policy) confers magical properties (e.g. originate a default route).  I think that can create fragile systems, where config and code are intermingled.  So instead of templating checking if e.g. the name of a particular policy is in the export policies for a peer, we should encourage the use of `config_context` of the routing policy in Peering Manager, e.g.:

```
{
    "id": 4,
    "display": "default-originate-IPv4",
    "name": "default-originate-IPv4",
    "slug": "default-originate-IPv4",
    "type": "export-policy",
    "weight": 0,
    "address_family": 4,
    "config_context": {
        "default-originate": true
    },

    /* or, if networks think that originating specific prefixes will be an important use-case, they could model as follows:
    "config_context": {
        "originate": [ "0.0.0.0/0" ]
    },
    */

    "comments": "",
    "tags": []
}
```

This effectively makes the `config_context` of the peering policies the "kitchen sink" for all things to do with what is carried within a BGP session.  However, is that the appropriate place for every other option we can think of for other models?

* BGP peering sessions (e.g. timers, confederations, etc)
* ASNs (e.g. this downstream customer has their own address space but also leases a /27 from us, and that needs to be added to their prefix list that is auto-generated from IRR data)
* Relationships (e.g. something to signify this peering is part of MPLS L3VPN rather than upstream/downstream/peer)

In #566 @gmazoyer asks:

> Add a config context object and allow it to be assigned per-object using generic foreign key?

What if we consider the Tag model as that per-object generic "config context"?  That would leave lots of flexibility for engineers to define whatever they like.  For example, they could define tag `slow-timers` which they can apply to BGP peers and groups, with a context of `{ "timers": { "keepalive": 60, "hold": 180 } }`

Of course this leaves the possibility of tagging an IX with a tag which is intended for direct and IX BGP sessions.  But that's no different to putting values in a routing policy's config_context which are not used by templates, so we don't lose anything.

## Caveats

Routing policies are an ordered list.  Tags are an unordered set.  Imagine these two hypothetical tags were applied to the same peer:

* tag **hold180**: `{ "timers": { "hold": 180, "b": [3] }, "a": [1] }`
* tag **keepalive60**: `{ "timers": { "keepalive": 60, "b": [3] }, "a": [2] }`

How does a template writer know what to do with the union of two config contexts?  Maybe the answer is to give the template-writer some "merge contexts" filters so they can decide.

### Salt-style merging

Lists concatenate, dicts merge.  Therefore **hold180** + **keepalive60** =

`{ "timers": { "keepalive": 60, "hold": 180, "b": [3, 4] }, "a": [1, 2] }`

### Netbox-style merging

Last one wins, so we need ways to sort to define "last".  Therefore **hold180** + **keepalive60** =

`{ "timers": { "keepalive": 60 }, "a": [2] }`

### Raise an exception

Enforce namespacing.  Therefore **hold180** + **keepalive60** =

"ERROR: 'timers' defined in tag 'hold180' cannot be redefined in 'keepalive60'"

### How to use Tag config contexts?

(these are draft proposals, not real examples of actual template code)

```jinja2
{% set tag_context = peer.tags|sort(attribute="slug")|merge_contexts_like_salt %}

neighbor {{ peer.address }} timers {{ tag_context.timers.keepalive or 10 }} {{ tag_context.timers.hold or 31 }}

/routing bgp peer add \
  {% if tag_context.passive %}passive=yes{% endif %} \
  remote-address={{ peer.address }}

{% if tag_context.override_local_as %}set protocols bgp {{ local_asn }} neighbor {{ peer.address }} local-as {{ tag_context.override_local_as }}{% endif %}
```

### Other Questions

* Do we add an "order" field to the Tag model, so there is a numeric field for sorting by?
* Or do we not overload the Tag model, and create a separate model for config contexts?
* And if we use a separate model, do we adopt a Netbox-style way of scoping them onto other models? (but there isn't the same hierarchy in Peering Manager as Netbox with sites/locations, tenants, etc)

**Reply by jamesditrapani** [reply]
@maznu This is a great write up and we appreciate the feedback. I actually don't mind the idea of making tags a config contextable model as it could work quite well. With regards to merging, I personally think raising an exception would be the ideal scenario. If you needed overlapping tags then its probably best to restructure your config context to be a bit more granular.

**Comment by mngan** [comment]
I can see uses of the Netbox-style scoping of config contexts. Netbox targets devices via various criteria. I can kind of see you wanting to target not just BGP Sessions, but perhaps BGP Groups/Internet Exchanges and Routers. So it could be a little different in that respect, but there is definitely a use case for the Netbox-style scoping. 

You may want BGP groups with specific tags or specific naming patterns to get one config context vs another. This avoids copy-pasta amongst a lot of similar BGP groups. 

You would will want to keep per object config contexts, but you would only use them in specific cases. In my case, I have some BGP sessions that need individual fine tuning. I haven't looked at how the config contexts work right now. I am guessing these are just available on the supported object types and there is no inheritance model at all. That would seem fine, if we could use the Netbox-style scoping to apply additional config contexts to objects. 

I am a little wary of the merge model. Netbox's merge model is very simple. It simply merges dicts at the top level rather than doing a nested merge, which is what Salt does, and more... For the simple merge, you really can't have complicated nested structures that you might want to deeply merge. It is only really good for adding new keys or completely overriding existing ones. The deep merge that Salt does usually works the way you want it to, except when it comes to lists. Sometimes you want it to overwrite, some times append. Salt will append by default, and it has a bunch of special keys to indicate how two dicts should be merged. So, I guess you could just provide the user with all the different contexts, and then let them merge the contexts via filters to suit their needs. 

As an aside, you could also leverage config contexts (or graphql query) from Netbox/Nautobot as well for the device. That could be another context that lives along side the Peering Manager ones if you have enabled Netbox integration. Some people might find that useful, but is really a separate thing from what we are talking about here.

**Reply by gmazoyer** [reply]
I should probably share what I have in mind as a developer for config context:

* Config context is in my opinion a way to enrich data which Peering Manager already have. So I do not intend to make the config context aware of any variables, that could be referenced with Jinja syntax or something.
* I believe NetBox has succeeded to implement config context. I like the fact that it can be defined at two levels, and I think it would be a good idea to bring that technique in. Regarding the merge procedure of the two contexts, Ansible also is a good example of what can be achieved with the [combine](https://docs.ansible.com/ansible/latest/user_guide/playbooks_filters.html#combining-hashes-dictionaries) filter. I was thinking about proposing different kind of merge solutions and let the user decide the one to use with a configuration setting.

My propositions to improve this feature are the following:

1. Rename the per object `config_context` field to `local_config_context`
2. Introduce a new `ConfigContext` object which can be assigned to types of objects (e.g. all IXPs)
3. Provide a merge technique to render the config context of an object based on the global and local contexts
4. Expose filters for the templating system to easily "grep" into a config context (e.g. `config_context_get_key('foo')`)
5. Extend config context to more objects (all of them except templating, tags, e-mail, contact related ones)

I probably forgot somethings but they'll come back during the development process.

**Reply by mngan** [reply]
This sounds good. 

How flexible do you see the assigning of the `ConfigContext` object? Are you thinking Netbox style with simple rules? Or just a global object assigned to each type of object? 

It would be nice to have a `jq` style filter to pull data from the context. Unfortunately, I only see one project `dq` that hasn't been maintained. 

Context objects on templates might serve a purpose of setting defaults or formatting parameters for a template? That way you keep that data out of the template. Not sure if others would find that useful or not.

**Reply by gmazoyer** [reply]
> How flexible do you see the assigning of the `ConfigContext` object? Are you thinking Netbox style with simple rules? Or just a global object assigned to each type of object?

I'm still not sure yet. At first I was thinking of a global object assigned to a type of object. My current local implementation does just that (screenshot attached). But I'm now thinking it might be not enough. So I was thinking about maybe doing something like "define a config context" (no object type or optional object type) and then "assign it to whatever objects you feel like with a generic foreign key".

> It would be nice to have a `jq` style filter to pull data from the context. Unfortunately, I only see one project `dq` that hasn't been maintained.

I did not know about `dq` but it could be nice to have. Maybe hard to implement though. So I'm pretty sure it'll be there from day one.

> Context objects on templates might serve a purpose of setting defaults or formatting parameters for a template? That way you keep that data out of the template. Not sure if others would find that useful or not.

Not a bad idea. If people find that useful it can be easily done.

<img width="934" alt="image" src="https://user-images.githubusercontent.com/736209/173800173-146d5dae-c2d2-470a-a1d0-52d0a6919faa.png">

**Comment by gmazoyer** [comment]
Implementation in progress https://github.com/peering-manager/peering-manager/pull/596
